genie-web/src/main/java/com/netflix/genie/web/services/impl/JobResolverServiceImpl.java

```diff
@@ -915,6 +925,62 @@ private Cluster resolveCluster(
         }
     }
 
+    /**
+     * Helper method to generate all the possible viable cluster criterion permutations for the given set of commands
+     * and the given job request. The resulting map will be each command to its associated priority ordered list of
+     * merged cluster criteria. The priority order is generated as follows:
+     * <pre>
+     * for (commandClusterCriterion : command.getClusterCriteria()) {
+     *     for (jobClusterCriterion : jobRequest.getClusterCriteria()) {
+     *         // merge
+     *     }
+     * }
+     * </pre>
+     *
+     * @param commands   The set of {@link Command}s whose cluster criteria should be evaluated
+     * @param jobRequest The {@link JobRequest} whose cluster criteria should be combined with the commands
+     * @return The resulting map of each command to their associated merged criterion list in priority order
+     */
+    private Map<Command, List<Criterion>> generateClusterCriteriaPermutations(
+        final Set<Command> commands,
+        final JobRequest jobRequest
+    ) {
+        final long start = System.nanoTime();
+        try {
+            final ImmutableMap.Builder<Command, List<Criterion>> mapBuilder = ImmutableMap.builder();
+            for (final Command command : commands) {
+                final ImmutableList.Builder<Criterion> listBuilder = ImmutableList.builder();
+                for (final Criterion commandClusterCriterion : command.getClusterCriteria()) {
+                    for (final Criterion jobClusterCriterion : jobRequest.getCriteria().getClusterCriteria()) {
+                        try {
+                            // Failing to merge the criteria is equivalent to a round-trip DB query that returns
+                            // zero results. This is an in memory optimization which also solves the need to implement
+                            // the db query as a join with a subquery.
+                            listBuilder.add(this.mergeCriteria(commandClusterCriterion, jobClusterCriterion));
+                        } catch (final IllegalArgumentException e) {
+                            log.debug(
```

Comment:
This doesn't deserve at least a `warn`?

Reply:
I don't think so in this case. The constructor for `Criterion` does indeed throw an exception for invalid configuration but at this point both the input criterion where valid so there's no real case where they can be invalid due to invalid data. The `IllegalArgumentException` is thrown by the helper method due to a case where the comment above the `add` line states because the query to the db can never succeed (e.g. conflicting `id` fields in the criterion can never succeed in matching anything)

It was a way to just signal back to this caller that the resulting criterion would be invalid. We could instead pass back `Optional` with empty representing that the two can't be merged if that makes the intent clearer.

Reply:
While I think it is safe to do this, it would probably be a better practice to log this as a warning like @mprimi suggests. That way if this is ever happening for an unexpected reason, we can clearly see what conflicted and whether that was correct or the result of a bug. I'd opt for visibility, though possibly at an info level rather than warn.